src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -78,124 +67,251 @@ public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirec
-                // to the real dotnet executable.
-                // This will work as often as just invoking dotnet from the command line, but we can be more confident in finding a dotnet executable by following
-                // https://github.com/dotnet/designs/blob/main/accepted/2021/install-location-per-architecture.md
-                // This can be done using the nethost library. We didn't do this previously, so I did not implement this extension.
-                foreach (string dir in Environment.GetEnvironmentVariable("PATH").Split(Path.PathSeparator))
+                var hostFxrRoot = Path.Combine(dotnetPath, "host", "fxr");
+                if (Directory.Exists(hostFxrRoot))
                 {
-                    string filePath = Path.Combine(dir, exeName);
-                    if (File.Exists(filePath))
+                    // Agreed to load hostfxr from the highest version
+                    var hostFxrAssemblyDirectory = Directory.GetDirectories(hostFxrRoot)
+                        .OrderByDescending(d => d)
+                        .FirstOrDefault();
+
+                    if (hostFxrAssemblyDirectory != null)
                     {
-                        if (!isWindows)
+                        Regex regex = new Regex("(lib)?hostfxr.dylib", RegexOptions.IgnoreCase);
+
+                        var matchingAssembly = Directory.GetFiles(hostFxrAssemblyDirectory)
+                            .Where(filePath => regex.IsMatch(Path.GetFileName(filePath)))
+                            .FirstOrDefault();
+
+                        if (matchingAssembly != null)
                         {
-                            filePath = realpath(filePath) ?? filePath;
-                            if (!File.Exists(filePath))
-                            {
-                                continue;
-                            }
-                        }
+                            var isHostFxrLoaded = AppDomain.CurrentDomain.GetAssemblies().Any(assembly =>
+                                string.Equals(assembly.GetName().Name, matchingAssembly, StringComparison.OrdinalIgnoreCase));
 
-                        dotnetPath = dir;
-                        break;
+                            if (!isHostFxrLoaded)
+                                Assembly.Load(matchingAssembly);
+                        }
                     }
                 }
             }
+        }
+
+        private static string? GetDotnetPathFromROOT()
+        {
+            // 32-bit architecture has (x86) suffix
+            string envVarName = (IntPtr.Size == 4) ? "DOTNET_ROOT(x86)" : "DOTNET_ROOT";
+            var dotnetPath = FindDotnetPathFromEnvVariable(envVarName);
+            
+            return dotnetPath;
+        }
 
-            if (dotnetPath is null)
+        private static string? GetDotnetPathFromDefaultPaths()
```

Comment:
This doesn't seem to be doing what the name suggests? I'd expected it to look in specific predefined directories.